### ISSUE TITLE
bindParams is not protected agains garbage collection

### ISSUE THREAD

**Issue by giniedp** [asifjalil/cli#12]
the `bindParam` methods seems to be unprotected agains the garbage collector.
https://github.com/asifjalil/cli/blob/85f2f56628027df278e75b3da6f4161c457ff147/stmt.go#L133

All pointers are created inside that method and passed to 
https://github.com/asifjalil/cli/blob/85f2f56628027df278e75b3da6f4161c457ff147/stmt.go#L248

After that no pointer is held by the go program.

We see unfrequent occurance of errors like 
```
CLI0111E Numeric value out of range. SQLSTATE=22003
```
(and other kinds). A retry of the same statement usually succseeds. However a series of large statements each with thousands of parameters like a batch insert have a greater chance to fail.

After investigating the issue we came up with a rough solution that makes the go program hold all bound params during the sql execution

```go
type bindParam struct {
    ctype   C.SQLSMALLINT
    sqltype C.SQLSMALLINT
    decimal C.SQLSMALLINT
    size    C.SQLULEN
    buflen  C.SQLLEN
    plen    *C.SQLLEN
    buf     unsafe.Pointer
}

func (s *stmt) bindParam(idx int, v driver.Value) (*bindParam, error) {
    bindParam := &bindParam{}
    // ...
    return bindParam, nil
}

func (s *stmt) sqlexec(ctx context.Context, args []driver.Value) error {
    // ...  
    bPList := make([]*bindParam, 0, len(args))
    // bind values to parameters
    for i, a := range args {
        // ...
        bPList = append(bPList, bP)
    }
    return nil
}
```

This ofc. can be improved. `plen` and `buf` are the ones that actually matter.

Looking at the implementation of phiggins/db2cli they have used a similar strategy
https://bitbucket.org/phiggins/db2cli/src/347a77fd71abb55325ebea9ca6c2a71c614e9a21/param.go#lines-22

**Comment by asifjalil** [comment]
@giniedp thanks for reporting the issue. I will take a look.

**Comment by giniedp** [comment]
to reproduce the faulty behavior just call manually `runtime.GC()` after the parameters are bound

**Comment by asifjalil** [comment]
@giniedp I updated the driver based on your suggestions. You are right: we only need to keep the reference for `plen` and `buf` so Go GC doesn't remove them between SQLBindParameter and SQLExecute call.  Please take a look at the commit when you get a chance.

I am not sure how to test this using the `database/sql` interface. If you have any suggestion please send them my way. 

p.s. thanks for finding this bug.

**Comment by giniedp** [comment]
Thank you for the fix. It looks good to me. 
I think it can not be tested reliably through the `database/sql` interface since for that `runtime.GC()` must be called between `s.bindParam` and `C.SQLExecute`. The only idea that comes to my mind is to execute a long running query inside a go routine and calling `runtime.GC()` right after that. However the timing is unpredictable hence its unreliable.

**Comment by shrek** [comment]
Thanks @giniedp for this find, and to @asifjalil for the fix.

We have been investigating a rare but nasty data-corruption in structures read from db-tables, and we suspected the garbage collector, and this issue is most likely what we are facing as well.

**Comment by Manaswi94** [comment]
@giniedp Very nice finding and thank you @asifjalil for the fix.

@asifjalil I see PR for this issue is merged. When can we expect a new release ?

**Comment by asifjalil** [comment]
@Manaswi94 I want to refactor the PR a bit before I create a new release. I already have a param struct for inout stored procedure parameters. I want to merge the new param type with that. I have done the refactor already but just double checking to make sure the change makes sense. 

I will make a new release shortly.

**Comment by asifjalil** [comment]
@giniedp thanks again for finding the root cause.

FYI.  I just released v1.2.6 with the fix.
